tox.ini

```diff
@@ -10,14 +10,19 @@ skip_missing_interpreters = true
 [testenv]
 basepython = python3
 # List common dependencies for Python interpreters here:
+# note that we dropped support for ansible 2.6, and ansible
+# 2.7 dropped support for py26 - so we only do ansible
+# related testing on py27 and later
 deps =
+    py{27,36,37,38}: ansible
```

Comment:
This seems to be a new dependency since it is not removed from the py26 env (it does not seem to be specified, there).

Reply:
ansible does not work with python 2.6
ansible is needed for pytest

Reply:
Are you sure? https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#minimum-version-of-python-3-x-and-python-2-x "Module-side, we support Python 3.5 or greater and Python 2.6 or greater." And this seems to be used for module code testing, not running Ansible itself.

Reply:
(maybe I don't understand why Ansible is needed for pytest.)

Reply:
> Are you sure? https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#minimum-version-of-python-3-x-and-python-2-x "Module-side, we support Python 3.5 or greater and Python 2.6 or greater." And this seems to be used for module code testing, not running Ansible itself.

I tried it - it doesn't work - I tried installing the latest ansible 2.7.x in a py26 tox venv and it complained about the version of python.  Please try it.  NOTE that I understand what the ansible documentation says, I'm suggesting that it may be incorrect.

Reply:
> (maybe I don't understand why Ansible is needed for pytest.)

in lsr, pytest is (almost?) always used to do unit testing for module code, which means the unit test code needs ansible.

Reply:
when did it complain? I would expect it to complain when you try to run the `ansible` command, bit not when you just import the required Python helper packages in a module. Otherwise our roles would not work against RHEL 6.

Reply:
It complained when I tried to `pip install ansible` in the tox py26 venv.  Again, please don't take my word for it, please try it for yourself, perhaps I did something wrong.

Reply:
Usually, the roles only import ansible because the module_utils are in the ansible namespace and to create an AnsibleModule object. As long as we need to support Python 2.6 for CentOS/RHEL 6, we should also tests against 2.6. So if the unit tests require ansible to test more than just the AnsibleModule object creation, this needs to be addressed or proper testing in 2.6. This might not the right place to discuss this because the unit tests in the network role do not require ansible for their tests.

Reply:
Ah, it is already the installation via pip that fails. I see that it would be a problem, yes.
@tyll but how to import module_utils if ansible (which contains it) can't be installed via pip?

Reply:
The tests mock the ansible module: https://github.com/linux-system-roles/network/blob/master/tests/unit/test_nm_provider.py#L17

Reply:
in `kernel_settings` and `storage` and possibly other roles, `ansible` is used directly in the unit tests.  I'm not sure if `ansible` is needed by `pylint` or `flake` tests.
If it is going to be a problem to have `ansible` always installed for every python environment, then I'll figure out a way to refactor this, to make the `ansible` dependency per-tox environment, or even a per-repo dependency installed via some sort of custom requirements file.

Reply:
Do the kernel_settings and storage role target CentOS 6, too, or only newer versions? Then it would not matter for them. But the reasoning in the comment does not make sense for other roles that also target CentOS6/Python 2.6.

Reply:
I'm coming up with a much more comprehensive way to specify that ansible is a dependency for pytests for different platforms, which also takes into consideration that installing ansible in a tox venv on 2.6 has additional requirements, both on the ubuntu platform side in the form of additional C library dependencies, and on the `pip` side in the form of older versions of some dependencies.  This will give the role developer the flexibility to decide if `ansible` is a dependency for the pytest tests, and which version of ansible to use.

Reply:
> The tests mock the ansible module: https://github.com/linux-system-roles/network/blob/master/tests/unit/test_nm_provider.py#L17

Hmm, is this approach going to be scalable and maintainable if used for all the roles that have embedded modules?

Reply:
> Hmm, is this approach going to be scalable and maintainable if used for all the roles that have embedded modules?

My approach is to give unit tests for modules an actual run-time environment that is as close to a production environment as possible, with a real ansible with the role module_utils/ code in the real `ansible.module_utils.*`module namespace, so that no `sys.path` munging, or module `Mock`-ing, is necessary.

That being said, I think the approach that `network` takes might be useful for certain unit test cases where you want to test a single method or small group of methods, and nothing else, and find that dragging in all of ansible is an impediment.

But I still think that unit testing with a real ansible is a good thing that we should do by default.